Source/Base/Base/Math/Spline.cpp

```diff
@@ -0,0 +1,261 @@
+    }
+
+    void Spline::AddPoint(const SplinePoint& newPoint, i32 position)
+    {
+        if (position < 0)
+        {
+            _points.emplace_back(newPoint);
+        }
+        else
+        {
+            _points.insert(_points.begin() + position, newPoint);
+        }
+
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::RemovePoint(u32 position)
+    {
+        if (position >= _points.size())
+            return;
+
+        std::vector<SplinePoint>::iterator it = _points.begin();
+        std::advance(it, position);
+        _points.erase(it);
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::SetStep(u32 newStep)
+    {
+        newStep = Math::Max((i32)newStep, 1);
+
+        if (_step == newStep)
+            return;
+
+        _step = newStep;
+        _storage.SetStep(newStep);
+    }
+
+    void Spline::SetType(InterpolationType newType)
+    {
+        if (newType < InterpolationType::None || newType >= InterpolationType::MAX_TYPE || _type == newType)
+            return;
+
+        _type = newType;
+        _storage.MarkAsDirty();
+    }
+
+    void Spline::SetAlpha(f32 newAlpha)
+    {
+        _alpha = Math::Clamp(newAlpha, 0.f, 1.f);
+        _storage.MarkAsDirty();
+    }
+
+    bool Spline::DoInterpolation()
+    {
+        if (!_storage.IsDirty())
+            return false;
+
+        // need to have minimum 2 points
+        if (_points.size() < 2 || _step == 0)
+            return false;
+
+        _storage.Clear();
+        f32 tStep = (1.f) / static_cast<f32>(_step);
+        if (IsInterpolatedWithControl())
+        {
+            for (i32 n = 0; n < _points.size() - 1; n++)
+            {
+                SplinePoint currentPoint = _points[n];
+                SplinePoint nextPoint = _points[n + 1];
+
+                std::vector<vec3> currentPortion;
+                currentPortion.reserve(_step + 1);
+
+                for (u32 i = 0; i < _step + 1; i ++)
+                {
+                    f32 t = tStep * static_cast<f32>(i);
+                    currentPortion.emplace_back(InterpolateControl(t, currentPoint, nextPoint));
+                }
+
+                _storage.AddPortion(currentPortion);
+            }
+        }
+        else
+        {
+            vec3* ptr = GetPtrPoints();
+            for (i32 n = 0; n < (_points.size() * 3) - 4; n++)
+            {
+                std::vector<vec3> currentPortion;
+                currentPortion.reserve(_step + 1);
+
+                for (u32 i = 0; i < _step + 1; i++)
+                {
+                    f32 t = tStep * static_cast<f32>(i);
+                    currentPortion.emplace_back(InterpolateCombined(t, ptr, n, _alpha));
+                }
+
+                _storage.AddPortion(currentPortion);
+            }
+        }
+
+        return true;
+    }
+
+    vec3* Spline::GetPtrPoints()
+    {
+        if (_points.empty())
+            return nullptr;
+
+        vec3* ptr = new vec3[_points.size() * 3];
```

Comment:
Probably makes more sense to let something else provide you a container, and then you can provide the size of the data. This is asking to become a memory leak